### PR TITLE
MNT Update dependencies

### PR DESCRIPTION
**Pull request by fcharras**
Update Dockerfile, patch around inline optimization issues, and hack around drop of atomics on CPU to keep a working CI (by pinning CI on an old working version).

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -62,35 +62,34 @@ ARG ONEAPI_INSTALL_DIR=/opt/intel/oneapi
...
 ARG DPCPP_LLVM_SPIRV_GIT_URL=https://github.com/IntelPython/dpcpp-llvm-spirv/
 
-
 # Version of other python packages explicitly installed either within the
 # build environment or the runtime environment.
 
 ARG NUMPY_VERSION="<1.25"
 ARG NUMBA_VERSION=""
-ARG CYTHON_VERSION="<3"
+ARG CYTHON_VERSION=""
```
Comment (ogrisel) [inline docker/Dockerfile:92]: If Cython is used to build scikit-learn, note that there are a few performance regressions with Cython 3. So we might want to stick to the latest Cython 2 for the time being.

**Review by ogrisel — APPROVED** [review]
If it works, LGTM :)

**Comment by ogrisel** [comment]
Apparently it doesn't work:

```
WARNING: The directory '/github/home/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
WARNING: Requirement '/opt/numba_dpex/cpu-stable/numba_dpex*.whl' looks like a filename, but the file does not exist
ERROR: numba_dpex*.whl is not a valid wheel filename.
```

**Comment by fcharras** [comment]
It's normal that the test don't pass yet, we must merge first so that the runner build and update the image to our repo, and re-run the pipelines after that. I'll perform a few final tests locally before.

File: docker/Dockerfile (hunk 1 above)
Comment (fcharras) [reply docker/Dockerfile:92]: cython is not installed in the final image, so it's not related to the cython one can use to build sklearn later on in a container. Most recent dpctl has merged work that makes it compatible with cython>=3 so I think we can keep that.

**Pull request merged**